src/D2L.CodeStyle.Analyzers/Immutability/ImmutableAttributeConsistencyChecker.cs

```diff
@@ -86,52 +86,88 @@ CancellationToken cancellationToken
 			ImmutableTypeInfo baseTypeInfo,
 			CancellationToken cancellationToken
 		) {
-			switch( baseTypeInfo.Kind ) {
-				case ImmutableTypeKind.None:
+			switch( ((baseTypeInfo.Kind, baseTypeInfo.IsConditional), (typeInfo.Kind, typeInfo.IsConditional)) ) {
+				default:
+					throw new NotImplementedException();
+
+				case ((ImmutableTypeKind.None, _), _):
+				case ((ImmutableTypeKind.Instance, _), _):
 					return;
 
-				case ImmutableTypeKind.Instance:
+				case ((ImmutableTypeKind.Total, _), (ImmutableTypeKind.None, _)):
+				case ((ImmutableTypeKind.Total, _), (ImmutableTypeKind.Instance, _)):
+					RaiseMissingAttribute();
 					return;
 
-				case ImmutableTypeKind.Total:
-					if( typeInfo.Kind == ImmutableTypeKind.Total ) {
-						break;
-					}
+				// If the implementing type is [Immutable] then the conditionality of the
+				// base type doesn't matter (for this diagnostic)
+				case ((ImmutableTypeKind.Total, _), (ImmutableTypeKind.Total, false)):
+					return;
 
-					// The docs say the only things that can have BaseType == null
-					// are interfaces, System.Object itself (won't come up in our
-					// analysis because (1) it !hasTheImmutableAttribute (2) you
-					// can't explicitly list it as a base class anyway) and
-					// pointer types (the base value type probably also doesn't
-					// have it.)
-					bool isInterface =
-						baseTypeInfo.Type.BaseType == null
-						&& baseTypeInfo.Type.SpecialType != SpecialType.System_Object
-						&& baseTypeInfo.Type.SpecialType != SpecialType.System_ValueType
-						&& baseTypeInfo.Type.Kind != SymbolKind.PointerType;
-
-					(TypeDeclarationSyntax syntax, _) = typeInfo.Type.ExpensiveGetSyntaxImplementingType(
-						baseTypeOrInterface: baseTypeInfo.Type,
-						compilation: m_compilation,
-						cancellationToken
-					);
-
-					m_diagnosticSink(
-						Diagnostic.Create(
-							Diagnostics.MissingTransitiveImmutableAttribute,
-							syntax.Identifier.GetLocation(),
-							properties: FixArgs,
-							typeInfo.Type.GetFullTypeName(),
-							baseTypeInfo.IsConditional ? " (or [ConditionallyImmutable])" : "",
-							isInterface ? "interface" : "base class",
-							baseTypeInfo.Type.GetFullTypeName()
-						)
-					);
-
-					break;
+				// If the base type is [Immutable] then the implementing type should be as well
+				case ((ImmutableTypeKind.Total, false), (ImmutableTypeKind.Total, true)):
+					RaiseMissingAttribute();

```

Comment:
```suggestion
					// TODO: error message could be improved to directly explain that we need [Immutable]
					RaiseMissingAttribute();
```

Reply:
I don't think it's urgent to actually implement but I think it'd be helpful to write down what we talked about in a comment